src/Common/src/CoreLib/System/MemoryExtensions.cs

```diff
@@ -186,11 +186,12 @@ public static bool IsWhiteSpace(this ReadOnlySpan<char> span)
         public static int IndexOf<T>(this Span<T> span, T value)
             where T : IEquatable<T>
         {
-            if (typeof(T) == typeof(byte))
+            if (IsTypeNumeric<T>(out int size))
```

Comment:
With `typeof(T) == typeof(byte)`, the JIT will avoid needing to generate any code for the other branch.  Is it able to with this pattern as well?

Reply:
Yes, the JIT is only generating the necessary code and treating size as a constant.

For byte:
![image](https://user-images.githubusercontent.com/6527137/37442689-f496d942-27c4-11e8-8eb2-fbf43bb6ac50.png)

For int:
![image](https://user-images.githubusercontent.com/6527137/37442709-0c2db814-27c5-11e8-91b5-22a8ebe29071.png)

Reply:
What does it look before/after for `Span<string>` ?

Reply:
It is bad for `Span<string>`. Before, all checks are avoided. Now, the IsTypeNumeric method is called with all the checks.

![image](https://user-images.githubusercontent.com/6527137/37445081-dae3da3c-27d2-11e8-88da-0318fe7a78e7.png)

Reply:
> I think if you add in the `default(T) == null` check for ref types as an explicit case it should work.

Andy's suggestion worked well:

![image](https://user-images.githubusercontent.com/6527137/37496189-48372742-286e-11e8-8fa0-74722227ae8f.png)

Reply:
It is definitely better, but there is still extra left over cruft. Should we rather look into generating the code using template, like we do for `Vector<T>` ?

Reply:
> Should we rather look into generating the code using template

Seems excessive for this especially since the code is identical for most of the types. Alternatively, we could just remove the IsTypeComparableAsBytes method and duplicate the code within the 6 call sites. Would a few extra instructions matter here (specifically for reference types like string)? There isn't any extra overhead for types like byte/int.

Reply:
https://github.com/dotnet/corefx/pull/28080#discussion_r174667868 showed extra instructions even for non-reference case.

Also the real code may suffer from this more than microbechmarks. The JIT has a fixed limits on the amount of code that it optimizes or number of local variables that it optimizes. All the extra complexity counts against these limits.

I would also make the AggressiveInlining code as streamlined as possible. In this case, it means duplicating the code withing the 6 call sites.

Reply:
>  showed extra instructions even for non-reference case.

The extra instructions in the byte case were a copy/paste mistake. Just like int, the byte case doesn't have overhead:
![image](https://user-images.githubusercontent.com/6527137/37497012-03fab486-2873-11e8-877f-a219816f995e.png)

> Also the real code may suffer from this more than microbechmarks. The JIT has a fixed limits on the amount of code that it optimizes or number of local variables that it optimizes. All the extra complexity counts against these limits.
I would always make the AggressiveInlining code as streamlined as possible. In this case, I think it means duplicating the code within the 6 call sites.

OK, I will duplicate the code then.

Reply:
> The JIT has a fixed limits on the amount of code that it optimizes or number of local variables that it optimizes. All the extra complexity counts against these limits. I would always make the AggressiveInlining code as streamlined as possible. In this case, I think it means duplicating the code within the 6 call sites.

> OK, I will duplicate the code then.

On second thought, I am not sure if this much duplication is worth it, given it would only add a few instructions specific to reference types and the code becomes long and cumbersome. If the JIT, in some cases, is unable to optimize this, then how will duplicating the code within the method help avoid that? I am not sure how we are reducing the complexity here. @jkotas, thoughts? Are there other benefits than having one less method that is marked with AggressiveInlining?

Reply:
If you inline the code manually, the JIT turns a lot of things into constants right away. The dead code can be pruned quickly, and the trees created are generally simple. If you let the JIT to do the inlining, there are complex trees created first and that the JIT needs to work hard to simplify. It has likely negative impact on both throughput; and on the code quality. I think you should be able to see the impact on code quality if you just manually unroll the microbenchmark to calls these methods say 50x. This should be enough to hit the JITs thresholds on too complex code. Measure the performance with both with and without manual inlining.

@AndyAyersMS Do you have an opinion whether it is better to inline manually or whether it is better to have the nested aggressively inlined methods here?

Reply:
> I think you should be able to see the impact on code quality if you manually unroll the microbenchmark to calls these methods say 50x. This should be enough to hit the JITs thresholds on too complex code. Measure the performance with both with and without manual inlining.

I call the method ~50x times (`result &= StartsWith(...)`) within each iteration of the benchmark. I wasn't able to see a performance difference:
![image](https://user-images.githubusercontent.com/6527137/37501750-e0cb1e50-288c-11e8-9e30-74cfbd505d06.png)

![image](https://user-images.githubusercontent.com/6527137/37502028-26992d5e-288e-11e8-93a6-9fa38678ca7f.png)

Reply:
The jit has gotten better at early (importer) pruning of dead code so often times we only import a slice of a method. We're also more aggressive about forwarding values into inline bodies and out of returns. 

So I think the nested aggressive inline is ok here from the jit's standpoint, and it probably makes the code more readable/maintainable.